source/slang/slang-state-serialize.cpp

```diff
@@ -65,21 +73,23 @@ struct StoreContext
     typedef StateSerializeUtil::SourceFileState SourceFileState;
     typedef StateSerializeUtil::PathInfoState PathInfoState;
 
-    StoreContext(RelativeContainer* container)
+    StoreContext(OffsetContainer* container)
     {
         m_container = container;
     }
 
-    Safe32Ptr<FileState> findFile(const String& uniqueIdentity)
```

Comment:
It seems like you could have kept the `Safe32Ptr` abstraction around, such that it recorded both the `OffsetContainer` and the offset, and then when dereferenced always went back to the container to get the base address. That would retain the convenience of the original way this code was written.

I guess what it doesn't get you is abstraction over the save-vs-load cases, since in the serialization case you have an `OffsetContainer` while in the loading case you just have a flat buffer, but that seems to be what `OffsetBase` is abstracting over.

Am I missing something?

Reply:
I could have kept it or something like it, but just having one type simplified the concepts. True now you need to always indirect through the 'base' when you didn't before. Whilst more verbose it makes clearer where the 'rawness' conversion happens, and thats important because you need to be pretty careful around raw pointers on construction.

Yes OffsetBase gives the most minimal functionality such that reading works. And OffsetContainer is for writing. I actually mode OffsetBase a base class of OffsetContainer in a fix since this review.